specs/AllowHostInputProcessing.md

```diff
@@ -0,0 +1,101 @@
+
+In order to solve this type of issue, we provide an api which allows user inputs pass through the browser to the host app.
+
+# Description
+`AllowHostInputProcessing` allows user input messages to pass through the browser window to be received by an app process window.
+
+The messages can be received by Win32 API ::GetMessage() or ::PeekMessage(). This provides the host app a chance to handle the message before dispatching to the WebView2 HWND.
+
+If the host app does not handle input, it is forwarded to the browser process on the user's behalf. This API does not introduce any requirement for the developer to forward all input as is the case with visual hosting. This API should not be used with visual hosting, and has no effect when using  CreateCoreWebView2CompositionControllerWithOptions to create the controller.
+
+# Examples
+## Win32 C++
+```cpp
+HRESULT AppWindow::CreateControllerWithInputPassthrough()
+{
+      //! [CreateControllerWithOptions]
+    auto webViewEnvironment10 = m_webViewEnvironment.try_query<ICoreWebView2Environment10>();
+    if (!webViewEnvironment10)
+    {
+        FeatureNotAvailable();
+        return S_OK;
+    }
+
+    wil::com_ptr<ICoreWebView2ControllerOptions> options;
+    CHECK_FAILURE(webViewEnvironment4->CreateCoreWebView2ControllerOptions(options.GetAddressOf()));
+
+    wil::com_ptr<ICoreWebView2ControllerOptions3> webView2ControllerOptions3;
+    if (SUCCEEDED(options->QueryInterface(IID_PPV_ARGS(&webView2ControllerOptions3))))
+    {
+        CHECK_FAILURE(webView2ControllerOptions3->put_AllowHostInputProcessing(TRUE));
+    }
+
+    CHECK_FAILURE(webViewEnvironment10->CreateCoreWebView2ControllerWithOptions(
+        m_mainWindow, options.get(),
+        Callback<ICoreWebView2CreateCoreWebView2ControllerCompletedHandler>(
+            this, &AppWindow::OnCreateCoreWebView2ControllerCompleted)
+            .Get()));
+
+    return S_OK;
+}
+```
+
+### .NET, WinRT
+```c#
+CoreWebView2Environment _webViewEnvironment;
+public CreateWebView2Controller(IntPtr parentWindow)
+{
+    CoreWebView2ControllerOptions controllerOptions = new CoreWebView2ControllerOptions();
+    controllerOptions.AllowHostInputProcessing = true;
+
+    CoreWebView2Controller controller = null;
+
+    controller = await _webViewEnvironment.CreateCoreWebView2ControllerAsync(parentWindow, controllerOptions);
+
+    //...
+}
+```
+
+# API Details
+## Win32 C++
+```IDL
+interface ICoreWebView2StagingControllerOptions;
+
+interface ICoreWebView2StagingControllerOptions : IUnknown {
+  /// `AllowHostInputProcessing` property is to enable/disable input passing through
+  /// the app before being delivered to the WebView2. This property is only applicable
+  /// to controllers created with `CoreWebView2Environment.CreateCoreWebView2ControllerAsync` and not
+  /// composition controllers created with `CoreWebView2Environment.CreateCoreWebView2CompositionControllerAsync`.
```

Comment:
What happens if I set it to TRUE but call CreateCoreWebView2CompositionControllerAsync?

Option 1: Flag is ignored.
Option 2; CreateCoreWebView2CompositionControllerAsync fails.

I think option 2 is better, because we failed to fulfill the app's demands.

Reply:
I agree Option 2 is standardized. But actually CreateCoreWebView2CompositionController, or visual hosting, also allows inputs to be sent to the host, and it then require developer to call APIs to send the input to browser. If I set it to FALSE, it looks like "I don't want it to be received in the host, just directly deliver it to browser", and actually we don't fulfill this demand. I did not find an appropriate situation to fail this API, so I simply made it has no effect in this situation.